components/pd_client/src/util.rs

```diff
@@ -456,43 +545,87 @@ impl PdConnector {
+        let client_urls = peer.get_client_urls();
+        for ep in client_urls {
+            match self.connect(ep.as_str()).await {
+                Ok((client, address, _)) => {
+                    info!("connected to PD member"; "endpoints" => ep);
+                    return Ok((Some((client, address)), false));
+                }
+                Err(Error::Grpc(RpcFailure(RpcStatus { status, details: _ }))) => {
+                    if status == RpcStatusCode::UNAVAILABLE
+                        || status == RpcStatusCode::DEADLINE_EXCEEDED
+                    {
+                        error!("failed to connect to PD member"; "endpoints" => ep, "status_code" => ?status);
+                        network_fail_num += 1;
+                    }
+                }
+                Err(e) => {
+                    error!("failed to connect to PD member"; "endpoints" => ep, "error" => ?e);
+                }
             }
         }
-        None
+        let url_num = client_urls.len();
+        if url_num != 0 && url_num == network_fail_num {
+            has_network_error = true;
+        }
+        Ok((None, has_network_error))
     }
 
     pub async fn reconnect_leader(
         &self,
-        previous: &GetMembersResponse,
-        force: bool,
-    ) -> Result<Option<StubPair>> {
-        let resp = self.load_members(&previous).await;
+        leader: &Member,
+    ) -> Result<(Option<(PdClientStub, String)>, bool)> {
+        fail_point!("connect_leader", |_| Ok((None, true)));
+        let mut retry_times = MAX_RETRY_TIMES;
 
-        // Then try to connect the PD cluster leader.
-        match resp {
-            Some(resp) => {
-                if !force && resp == *previous {
-                    return Ok(None);
-                }
-                match self.connect_member(resp.get_leader()).await {
-                    Some(client) => Ok(Some((client, resp))),
-                    None => Err(box_err!("failed to connect to {:?}", resp.get_leader())),
+        // Try to connect the PD cluster leader.
+        loop {
+            let (res, has_network_err) = self.connect_member(leader).await?;
+            match res {
+                Some((client, address)) => return Ok((Some((client, address)), has_network_err)),
+                None => {
+                    if has_network_err && retry_times > 0 {
+                        let _ = GLOBAL_TIMER_HANDLE
+                            .delay(Instant::now() + Duration::from_secs(RETRY_INTERVAL_SEC))
+                            .compat()
+                            .await;
+                        retry_times -= 1;
+                        continue;
+                    }
+                    return Ok((None, has_network_err));
                 }
             }
-            None => Err(box_err!(
-                "failed to connect to {:?}",
-                previous.get_members()
-            )),
         }
     }
+
+    pub async fn reconnect_followers(
```

Comment:
We can also record a `TODO` to connect the PD which locates in the same zone first.